src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -437,86 +475,56 @@ protected internal void OnSolutionReloaded(SolutionInfo reloadedSolutionInfo)
         /// </summary>
         protected internal void OnSolutionRemoved()
         {
-            using (_serializationLock.DisposableWait())
-            {
-                var oldSolution = this.CurrentSolution;
-
-                this.ClearSolutionData();
-
-                // reset to new empty solution
-                this.SetCurrentSolution(this.CreateSolution(SolutionId.CreateNewId()));
-
-                this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.SolutionRemoved, oldSolution, this.CurrentSolution);
-            }
+            this.SetCurrentSolution(
+                _ => this.CreateSolution(SolutionId.CreateNewId()),
+                WorkspaceChangeKind.SolutionRemoved,
+                onBeforeUpdate: (_, _) => this.ClearSolutionData());
```

Comment:
more complex case.  not only is a new solution being created, and an event being issued, but we also need to mutate workspace state prior to setting the solution.  So onBeforeUpdate handles that here.